### PR TITLE
ocm: add klusterlet addon config and unit tests 

### PR DESCRIPTION
**Pull request by klaskosk**
Adds the KlusterletAddonConfig to the existing ocm package along with unit tests for the new resource.

### CODE REVIEW

File: pkg/kac/kac.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+		return nil, fmt.Errorf("klusterletAddonConfig object %s does not exist in namespace %s", name, nsname)
+	}
+
+	builder.Definition = builder.Object
+
+	return builder, nil
+}
+
+// Exists checks whether the given KlusterletAddonConfig exists on the apiClient.
```
Comment (kononovn) [inline pkg/kac/kac.go:114]: on the cluster*

File: pkg/kac/kac.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+	}, klusterletAddonConfig)
+
+	if err == nil {
+		builder.Object = klusterletAddonConfig
+	}
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// Create makes a KlusterletAddonConfig on the apiClient if it does not already exist.
```
Comment (kononovn) [inline pkg/kac/kac.go:136]: on the cluster*

Comment (kononovn) [inline pkg/kac/kac.go:136]: on the cluster*

File: pkg/kac/kac.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	builder.Object = builder.Definition
+
+	return builder, err
+}
+
+// Update changes the existing KlusterletAddonConfig resource on the apiClient, falling back to deleting and recreating
```
Comment (kononovn) [inline pkg/kac/kac.go:159]: on the cluster*

File: pkg/kac/kac.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+func (builder *Builder) Delete() error {
+	if valid, err := builder.validate(); !valid {
+		return err
+	}
+
+	glog.V(100).Infof(
+		"Deleting KlusterletAddonConfig %s in namespace %s", builder.Definition.Name, builder.Definition.Namespace)
+
+	if !builder.Exists() {
+		builder.Object = nil
```
Comment (kononovn) [inline pkg/kac/kac.go:209]: Please add glog message

**Review by kononovn — CHANGES_REQUESTED** [review]
Looks good just few minor comments.

**Comment by klaskosk** [comment]
@kononovn Thanks for your review, I've updated the PR accordingly

File: pkg/kac/kac.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+	if valid, err := builder.validate(); !valid {
+		return err
+	}
+
+	glog.V(100).Infof(
+		"Deleting KlusterletAddonConfig %s in namespace %s", builder.Definition.Name, builder.Definition.Namespace)
+
+	if !builder.Exists() {
+		glog.V(100).Infof(
+			"KlusterletAddonConfig %s in namespace %s already does not exist",
```
Comment (kononovn) [inline pkg/kac/kac.go:210]: Please remove "already"

**Review by kononovn — CHANGES_REQUESTED** [review]
1 last comment to fix

**Review by kononovn — DISMISSED** [review]
/lgtm

**Comment by trewest** [comment]
@klaskosk Shouldn't this be in the ocm pkg?

**Comment by klaskosk** [comment]
@trewest I'm not too familiar with what counts as OCM or not, but if you think it would be better as part of OCM I can move it

**Comment by trewest** [comment]
> @trewest I'm not too familiar with what counts as OCM or not, but if you think it would be better as part of OCM I can move it

@klaskosk yeah I typically try to group resources by group as much as I can
```
$ oc explain klusterletaddonconfigs.agent.open-cluster-management.io | head -n 3
GROUP:      agent.open-cluster-management.io
KIND:       KlusterletAddonConfig
VERSION:    v1
$ oc explain managedcluster | head -n 3
GROUP:      cluster.open-cluster-management.io
KIND:       ManagedCluster
VERSION:    v1
```
If it isn't too much trouble, let's move it into the ocm pkg

**Comment by klaskosk** [comment]
@trewest Ah, thanks for the explanation. I've updated the PR to move it to the ocm package

File: pkg/ocm/kac.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+// KACBuilder provides a struct for the KlusterletAddonConfig resource containing a connection to the cluster and the
+// KlusterletAddonConfig definition.
+type KACBuilder struct {
+	// Definition of the KlusterletAddonConfig used to create the object.
+	Definition *kacv1.KlusterletAddonConfig
+	// Object of the KlusterletAddonConfig as it is on the cluster.
+	Object *kacv1.KlusterletAddonConfig
+	// apiClient used to interact with the cluster.
+	apiClient *clients.Settings
```
Comment (trewest) [inline pkg/ocm/kac.go:24]: I think this should be runtimeclient.Client

File: pkg/ocm/kac.go [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+	// Object of the KlusterletAddonConfig as it is on the cluster.
+	Object *kacv1.KlusterletAddonConfig
+	// apiClient used to interact with the cluster.
+	apiClient runtimeclient.Client
+	// errorMsg used to store latest error message from functions that do not return errors.
+	errorMsg string
+}
+
+// NewKACBuilder creates a new instance of a KlusterletAddonConfig builder.
+func NewKACBuilder(apiClient *clients.Settings, name, nsname string) *KACBuilder {
```
Comment (trewest) [inline pkg/ocm/kac.go:30]: @klaskosk Can you verify if there aren't any required fields when creating a KlusterletAddonConfig? I see these on my cluster:
- applicationManager
- certPolicyController
- policyController
- searchCollector

Comment (klaskosk) [reply pkg/ocm/kac.go:30]: @trewest Yes, those are the only four fields [not listed as optional](https://pkg.go.dev/github.com/stolostron/klusterlet-addon-controller@v0.0.0-20240606130554-01338045271a/pkg/apis/agent/v1#KlusterletAddonConfigSpec), but [each field in that struct is optional](https://pkg.go.dev/github.com/stolostron/klusterlet-addon-controller@v0.0.0-20240606130554-01338045271a/pkg/apis/agent/v1#KlusterletAddonAgentConfigSpec) so I think it's okay to just leave them as their zero values

**Review by trewest — APPROVED** [review]
/lgtm

**Pull request merged**